src/ucan.rs

```diff
@@ -0,0 +1,463 @@
+            Some(JsonValue::String(v)) if v == "0.8.1" => (),
+            _ => return Err(Error::MissingType),
+        }
+
+        if !payload.audience.starts_with("did:") {
+            return Err(Error::DIDURL);
+        }
+
+        // extract or deduce signing key
+        let key: JWK = match (
+            payload.issuer.get(..8),
+            &parts.header.jwk,
+            dereference(resolver, &payload.issuer, &Default::default())
+                .await
+                .1,
+        ) {
+            // did:pkh with and without fragment
+            (Some("did:pkh:"), Some(jwk), Content::Object(_) | Content::DIDDocument(_)) => {
+                match_key_with_did_pkh(&jwk, &payload.issuer)?;
+                jwk.clone()
+            }
+            // did:key without fragment
+            (Some("did:key:"), _, Content::DIDDocument(d)) => d
+                .verification_method
+                .iter()
+                .flatten()
+                .next()
+                .and_then(|v| match v {
+                    VerificationMethod::Map(vm) => Some(vm),
+                    _ => None,
+                })
+                .ok_or_else(|| Error::VerificationMethodMismatch)?
+                .get_jwk()?,
+            // general case, did with fragment
+            (Some(_), _, Content::Object(Resource::VerificationMethod(vm))) => vm.get_jwk()?,
+            _ => return Err(Error::VerificationMethodMismatch),
+        };
+
+        verify_bytes(
+            parts.header.algorithm,
+            &parts.signing_input,
+            &key,
+            &parts.signature,
+        )?;
+
+        let parents = try_join_all(
+            payload
+                .proof
+                .iter()
+                .map(|s| Self::decode_verify(s, resolver)),
+        )
+        .await?;
+
+        if parents
+            .iter()
+            .any(|p| p.ucan.payload.audience != payload.issuer)
+        {
+            return Err(Error::InvalidIssuer);
+        }
+
+        Ok(DecodedUcanTree {
+            // decode and verify parents
+            parents,
+            ucan: Ucan {
+                header: parts.header,
+                payload,
+                signature: parts.signature,
+            },
+        })
+    }
+
+    pub fn decode(jwt: &str) -> Result<Self, Error>
+    where
+        F: DeserializeOwned,
+        A: DeserializeOwned,
+    {
+        let parts = split_jws(jwt).and_then(|(h, p, s)| decode_jws_parts(h, p.as_bytes(), s))?;
+        let payload: Payload<F, A> = serde_json::from_slice(&parts.payload)?;
+        Ok(Self {
+            header: parts.header,
+            payload,
+            signature: parts.signature,
+        })
+    }
+
+    pub fn parents(&self) -> ParentIter {
+        self.payload.parents()
+    }
+}
+
+fn match_key_with_did_pkh(key: &JWK, did: &str) -> Result<(), Error> {
+    use std::str::FromStr;
+    Ok(crate::caip10::BlockchainAccountId::from_str(
```

Comment:
This is good but won't support uses of legacy/non-CAIP-10 `did:pkh` DIDs. Also, the part after the fragment isn't checked; could this cause a problem?
Instead, you could get the CAIP-10 value from the blockchainAccount id property of the verification method map - which the caller has access to (although if the without-fragment form of the issuer id is used, then the VM will have to be manually selected, like you are doing for `did:key`).

Reply:
should be addressed now I think

Reply:
230a7f64ac2f0c8e5b3740421ba599165cdff7dd fixes it; thanks.